internal/profiling/buffer.go

```diff
@@ -0,0 +1,270 @@
+// (see below), we need to divide N; since exponents of two are only divisible
+// by other exponents of two, we use floorCPUCount number of queuePairs within
+// each circularBuffer.
+//
+// Floor of the number of CPUs (and not the ceiling) was found to the be the
+// optimal number through experiments.
+func floorCPUCount() uint32 {
+	n := uint32(runtime.NumCPU())
+	for i := uint32(1 << 31); i >= 2; i >>= 1 {
+		if n&i > 0 {
+			return i
+		}
+	}
+
+	return 1
+}
+
+var numCircularBufferPairs = floorCPUCount()
+var numCircularBufferPairsMask = numCircularBufferPairs - 1
+
+// circularBuffer stores the Pushed elements in-memory. It uses a collection of
+// queuePairs to reduce contention on the acquired and written counters within
+// each queuePair; that is, at any given time, there may be floorCPUCount()
+// Pushes happening simultaneously.
+type circularBuffer struct {
+	mu sync.Mutex
+	qp []*queuePair
+	// qpn is an monotonically incrementing counter that's used to determine
+	// which queuePair a Push operation should to write to. This approach's
+	// performance was found to be better than writing to a random queue.
+	qpn uint32
+}
+
+var errInvalidCircularBufferSize = errors.New("buffer size is not an exponent of two")
+
+// Allocates a circular buffer of size size and returns a reference to the
+// struct. Only circular buffers of size 2^k are allowed (saves us from having
+// to do expensive modulo operations).
+func newCircularBuffer(size uint32) (*circularBuffer, error) {
+	if size&(size-1) != 0 {
+		return nil, errInvalidCircularBufferSize
+	}
+
+	cb := &circularBuffer{
+		qp: make([]*queuePair, numCircularBufferPairs),
+	}
+
+	for i := uint32(0); i < numCircularBufferPairs; i++ {
+		cb.qp[i] = newQueuePair(size / numCircularBufferPairs)
+	}
+
+	return cb, nil
+}
+
+// Pushes an element in to the circular buffer.
+func (cb *circularBuffer) Push(x interface{}) {
+	n := atomic.AddUint32(&cb.qpn, 1) & numCircularBufferPairsMask
+reloadq:
+	q := (*queue)(atomic.LoadPointer(&cb.qp[n].q))
+
+	acquired := atomic.AddUint32(&q.acquired, 1) - 1
+
+	if atomic.LoadUint32(&q.drainingPostCheck) > 0 {
+		// Between our q load and acquired increment, a drainer began execution
+		// and switched the queues. This is NOT okay because we don't know if
+		// acquired was incremented before or after the drainer's check for
+		// acquired == writer. And we can't find which without expensive
+		// operations, which we'd like to avoid. If the acquired increment was
+		// after, we cannot write to this buffer as the drainer's collection may
+		// have already started; we must write to the other queue.
+		//
+		// Reverse our increment and retry. Since there's no SubUint32 in atomic,
+		// ^uint32(0) is used to denote -1.
+		atomic.AddUint32(&q.acquired, ^uint32(0))
+		goto reloadq
+	}
+
+	// At this point, we're definitely writing to the right queue. That is, one
+	// of the following is true:
+	//   1. No drainer is in execution.
```

Comment:
A drainer is draining the other queue? (Essentially same as no drainer, though..)